File: android/src/main/kotlin/com/marqably/flutter_oewa/EventHelper.kt
package com.marqably.flutter_oewa

import android.util.Log
import de.infonline.lib.IOLAdvertisementEvent
import de.infonline.lib.IOLAdvertisementEvent.IOLAdvertisementEventType
import de.infonline.lib.IOLAudioEvent
import de.infonline.lib.IOLAudioEvent.IOLAudioEventType
import de.infonline.lib.IOLBackgroundTaskEvent
import de.infonline.lib.IOLBackgroundTaskEvent.IOLBackgroundTaskEventType
import de.infonline.lib.IOLCustomEvent
import de.infonline.lib.IOLDataEvent
import de.infonline.lib.IOLDataEvent.IOLDataEventType
import de.infonline.lib.IOLDeviceOrientationEvent
import de.infonline.lib.IOLDeviceOrientationEvent.IOLDeviceOrientationEventType
import de.infonline.lib.IOLDocumentEvent
import de.infonline.lib.IOLDocumentEvent.IOLDocumentEventType
import de.infonline.lib.IOLDownloadEvent
import de.infonline.lib.IOLDownloadEvent.IOLDownloadEventType
import de.infonline.lib.IOLEvent
import de.infonline.lib.IOLGameEvent
import de.infonline.lib.IOLGameEvent.IOLGameEventType
import de.infonline.lib.IOLGestureEvent
import de.infonline.lib.IOLGestureEvent.IOLGestureEventType
import de.infonline.lib.IOLHardwareButtonEvent
import de.infonline.lib.IOLHardwareButtonEvent.IOLHardwareButtonEventType
import de.infonline.lib.IOLIAPEvent
import de.infonline.lib.IOLIAPEvent.IOLIAPEventType
import de.infonline.lib.IOLLoginEvent
import de.infonline.lib.IOLLoginEvent.IOLLoginEventType
import de.infonline.lib.IOLOpenAppEvent
import de.infonline.lib.IOLOpenAppEvent.IOLOpenAppEventType
import de.infonline.lib.IOLPushEvent
import de.infonline.lib.IOLPushEvent.IOLPushEventType
import de.infonline.lib.IOLUploadEvent
import de.infonline.lib.IOLUploadEvent.IOLUploadEventType
import de.infonline.lib.IOLVideoEvent
import de.infonline.lib.IOLVideoEvent.IOLVideoEventType
import de.infonline.lib.IOLViewEvent
import de.infonline.lib.IOLViewEvent.IOLViewEventType

/**
 * Based on the EventHelper.java example code from the OEWA SDK Demo by Lars Heinze (20.11.17).
 */
object EventHelper {
    /**
     * Takes a combined identifier + state string (separated by ".") and returns the corresponding IOLEvent instance
     */
    fun getIOLEventFromString(
        eventname: String,
        category: String?,
        comment: String?,
        customParams: Map<String?, String?>?
    ): IOLEvent? {
        val eventSplit = eventname.split("\\.".toRegex(), limit = 2).toTypedArray()
        val eventIdentifier = eventSplit[0]
        val eventState = eventSplit[1]
        if (eventIdentifier == IOLAdvertisementEvent.eventIdentifier) {
            for (type in IOLAdvertisementEventType.values()) {
                if (type.state.equals(eventState, ignoreCase = true)) {
                    return IOLAdvertisementEvent(type, category, comment, customParams)
                }
            }
        } else if (eventIdentifier == IOLAudioEvent.eventIdentifier) {
            for (type in IOLAudioEventType.values()) {
                if (type.state.equals(eventState, ignoreCase = true)) {
                    return IOLAudioEvent(type, category, comment, customParams)
                }
            }
        } else if (eventIdentifier == IOLBackgroundTaskEvent.eventIdentifier) {
            for (type in IOLBackgroundTaskEventType.values()) {
                if (type.state.equals(eventState, ignoreCase = true)) {
                    return IOLBackgroundTaskEvent(type, category, comment, customParams)
                }
            }
        } else if (eventIdentifier == IOLDataEvent.eventIdentifier) {
            for (type in IOLDataEventType.values()) {
                if (type.state.equals(eventState, ignoreCase = true)) {
                    return IOLDataEvent(type, category, comment, customParams)
                }
            }
        } else if (eventIdentifier == IOLDeviceOrientationEvent.eventIdentifier) {
            for (type in IOLDeviceOrientationEventType.values()) {
                if (type.state.equals(eventState, ignoreCase = true)) {
                    return IOLDeviceOrientationEvent(type, category, comment, customParams)
                }
            }
        } else if (eventIdentifier == IOLDocumentEvent.eventIdentifier) {
            for (type in IOLDocumentEventType.values()) {
                if (type.state.equals(eventState, ignoreCase = true)) {
                    return IOLDocumentEvent(type, category, comment, customParams)
                }
            }
        } else if (eventIdentifier == IOLDownloadEvent.eventIdentifier) {
            for (type in IOLDownloadEventType.values()) {
                if (type.state.equals(eventState, ignoreCase = true)) {
                    return IOLDownloadEvent(type, category, comment, customParams)
                }
            }
        } else if (eventIdentifier == IOLGameEvent.eventIdentifier) {
            for (type in IOLGameEventType.values()) {
                if (type.state.equals(eventState, ignoreCase = true)) {
                    return IOLGameEvent(type, category, comment, customParams)
                }
            }
        } else if (eventIdentifier == IOLGestureEvent.eventIdentifier) {
            for (type in IOLGestureEventType.values()) {
                if (type.state.equals(eventState, ignoreCase = true)) {
                    return IOLGestureEvent(type, category, comment, customParams)
                }
            }
        } else if (eventIdentifier == IOLHardwareButtonEvent.eventIdentifier) {
            for (type in IOLHardwareButtonEventType.values()) {
                if (type.state.equals(eventState, ignoreCase = true)) {
                    return IOLHardwareButtonEvent(type, category, comment, customParams)
                }
            }
        } else if (eventIdentifier == IOLIAPEvent.eventIdentifier) {
            for (type in IOLIAPEventType.values()) {
                if (type.state.equals(eventState, ignoreCase = true)) {
                    return IOLIAPEvent(type, category, comment, customParams)
                }
            }
        } else if (eventIdentifier == IOLLoginEvent.eventIdentifier) {
            for (type in IOLLoginEventType.values()) {
                if (type.state.equals(eventState, ignoreCase = true)) {
                    return IOLLoginEvent(type, category, comment, customParams)
                }
            }
        } else if (eventIdentifier == IOLOpenAppEvent.eventIdentifier) {
            for (type in IOLOpenAppEventType.values()) {
                if (type.state.equals(eventState, ignoreCase = true)) {
                    return IOLOpenAppEvent(type, category, comment, customParams)
                }
            }
        } else if (eventIdentifier == IOLPushEvent.eventIdentifier) {
            for (type in IOLPushEventType.values()) {
                if (type.state.equals(eventState, ignoreCase = true)) {
                    return IOLPushEvent(type, category, comment, customParams)
                }
            }
        } else if (eventIdentifier == IOLUploadEvent.eventIdentifier) {
            for (type in IOLUploadEventType.values()) {
                if (type.state.equals(eventState, ignoreCase = true)) {
                    return IOLUploadEvent(type, category, comment, customParams)
                }
            }
        } else if (eventIdentifier == IOLViewEvent.eventIdentifier) {
            for (type in IOLViewEventType.values()) {
                if (type.state.equals(eventState, ignoreCase = true)) {
                    return IOLViewEvent(type, category, comment, customParams)
                }
            }
        } else if (eventIdentifier == IOLVideoEvent.eventIdentifier) {
            for (type in IOLVideoEventType.values()) {
                if (type.state.equals(eventState, ignoreCase = true)) {
                    return IOLVideoEvent(type, category, comment, customParams)
                }
            }
        } else if (eventIdentifier == IOLCustomEvent.eventIdentifier) {
            return IOLCustomEvent(eventState, category, comment, customParams)
        } else {
            Log.e("OEWA EventHelper", "Could not create OEWA event from identifier  Has to be one of the predefined types: advertisement, audio, backgroundTask, cstom, data, deviceOrientation, document, download, game, gesture, hardwareButton, iAP, login, openApp, push, upload, video, view")
        }

        Log.e("OEWA EventHelper", "unknown state for identifier: $eventname")
        return null
    }
}

File: ios/Classes/EventHelper.swift
import Foundation
import INFOnlineLibrary

class EventHelper {
    /**
     * A full map of all predefined event types and their allowed states
     */
    static var eventTypeStateMap = [
        "advertisement.open": IOLAdvertisementEventType.open,
        "advertisement.close": IOLAdvertisementEventType.close,
        "audio.play": IOLAudioEventType.play,
        "audio.pause": IOLAudioEventType.pause,
        "audio.stop": IOLAudioEventType.stop,
        "audio.next": IOLAudioEventType.next,
        "audio.previous": IOLAudioEventType.previous,
        "audio.replay": IOLAudioEventType.replay,
        "audio.seekBack": IOLAudioEventType.seekBack,
        "audio.seekForward": IOLAudioEventType.seekForward,
        "background_task.start": IOLBackgroundTaskEventType.start,
        "background_task.end": IOLBackgroundTaskEventType.end,
        "data.cancelled": IOLDataEventType.cancelled,
        "data.refresh": IOLDataEventType.refresh,
        "data.succeeded": IOLDataEventType.succeeded,
        "data.failed": IOLDataEventType.failed,
        "device_orientation.changed": IOLViewEventType.appeared, // this event type only has one option, so this is just here so we still get into the correct switch, but not type is needed
        "document.open": IOLDocumentEventType.open,
        "document.edit": IOLDocumentEventType.edit,
        "document.close": IOLDocumentEventType.close,
        "download.cancelled": IOLDownloadEventType.cancelled,
        "download.start": IOLDownloadEventType.start,
        "download.succeeded": IOLDownloadEventType.succeeded,
        "download.failed": IOLDownloadEventType.failed,
        "game.action": IOLGameEventType.action,
        "game.started": IOLGameEventType.started,
        "game.finished": IOLGameEventType.finished,
        "game.won": IOLGameEventType.won,
        "game.lost": IOLGameEventType.lost,
        "game.highscore": IOLGameEventType.newHighscore,
        "game.achievement": IOLGameEventType.newAchievement,
        "gesture.shake": IOLViewEventType.appeared, // this event type only has one option, so this is just here so we still get into the correct switch, but not type is needed
        "hardware_button.pushed": IOLViewEventType.appeared, // this event type only has one option, so this is just here so we still get into the correct switch, but not type is needed
        "iap.started": IOLIAPEventType.started,
        "iap.finished": IOLIAPEventType.finished,
        "iap.cancelled": IOLIAPEventType.cancelled,
        "login.succeeded": IOLLoginEventType.succeeded,
        "login.failed": IOLLoginEventType.failed,
        "login.logout": IOLLoginEventType.logout,
        "open_app.maps": IOLOpenAppEventType.maps,
        "open_app.other": IOLOpenAppEventType.other,
        "push.received": IOLViewEventType.appeared, // this event type only has one option, so this is just here so we still get into the correct switch, but not type is needed
        "upload.cancelled": IOLUploadEventType.cancelled,
        "upload.start": IOLUploadEventType.start,
        "upload.succeeded": IOLUploadEventType.succeeded,
        "upload.failed": IOLUploadEventType.failed,
        "video.play": IOLVideoEventType.play,
        "video.pause": IOLVideoEventType.pause,
        "video.stop": IOLVideoEventType.stop,
        "video.next": IOLVideoEventType.next,
        "video.previous": IOLVideoEventType.previous,
        "video.replay": IOLVideoEventType.replay,
        "video.seekBack": IOLVideoEventType.seekBack,
        "video.seekForward": IOLVideoEventType.seekForward,
        "view.appeared": IOLViewEventType.appeared,
        "view.refreshed": IOLViewEventType.refreshed,
        "view.disappeared": IOLViewEventType.disappeared
    ] as [String : Any];
    
  /**
   * Tries to find the best matching EventType from our event name string.
   * The event name string is supposed to be eventIdentifier + eventState separated by "." (i.e: "view.appeared")
   *
   * If no predefined and specific event type was found, we will return a custom event.
   */
  static func getIOLEventFromString(
    eventname: String,
    category: String?,
    comment: String?,
    customParams: [String: String]?
  ) -> IOLEvent? {
    let eventSplit = eventname.split(separator: ".", maxSplits: 1).map(String.init)
    let eventIdentifier = eventSplit[0]
      
    // if we got one of out predefined event types and states -> return specific event
      if (eventTypeStateMap[eventname] != nil) {
          let eventStateType = eventTypeStateMap[eventname];
          
          // try to find the best matching event to return
          switch(eventIdentifier) {
              case "advertisement": return IOLAdvertisementEvent(type: eventStateType as! IOLAdvertisementEventType, category: category, comment: comment, parameter: customParams);
              case "audio": return IOLAudioEvent(type: eventStateType as! IOLAudioEventType, category: category, comment: comment, parameter: customParams);
              case "background_task": return IOLBackgroundTaskEvent(type: eventStateType as! IOLBackgroundTaskEventType, category: category, comment: comment, parameter: customParams);
              case "data": return IOLDataEvent(type: eventStateType as! IOLDataEventType, category: category, comment: comment, parameter: customParams);
              case "device_orientation": return IOLDeviceOrientationEvent(category: category, comment: comment, parameter: customParams);
              case "document": return IOLDocumentEvent(type: eventStateType as! IOLDocumentEventType, category: category, comment: comment, parameter: customParams);
              case "download": return IOLDownloadEvent(type: eventStateType as! IOLDownloadEventType, category: category, comment: comment, parameter: customParams);
              case "game": return IOLGameEvent(type: eventStateType as! IOLGameEventType, category: category, comment: comment, parameter: customParams);
              case "gesture": return IOLGestureEvent(category: category, comment: comment, parameter: customParams);
              case "hardware_button": return IOLHardwareButtonEvent(category: category, comment: comment, parameter: customParams);
              case "iap": return IOLIAPEvent(type: eventStateType! as! IOLIAPEventType, category: category, comment: comment, parameter: customParams);
              case "login": return IOLLoginEvent(type: eventStateType as! IOLLoginEventType, category: category, comment: comment, parameter: customParams);
              case "open_app": return IOLOpenAppEvent(type: eventStateType as! IOLOpenAppEventType, category: category, comment: comment, parameter: customParams);
              case "push": return IOLPushEvent(category: category, comment: comment, parameter: customParams);
              case "upload": return IOLUploadEvent(type: eventStateType as! IOLUploadEventType, category: category, comment: comment, parameter: customParams);
              case "video": return IOLVideoEvent(type: eventStateType as! IOLVideoEventType, category: category, comment: comment, parameter: customParams);
              case "view": return IOLViewEvent(type: eventStateType as! IOLViewEventType, category: category, comment: comment, parameter: customParams);

              default: return IOLCustomEvent(name: eventname,  category: category, comment: comment, parameter: customParams);
          }
      }
      
    // otherwise return a custom event
     return IOLCustomEvent(name: eventname,  category: category, comment: comment, parameter: customParams);
  }
}
